### PR TITLE
[Feature] add visibility prediction head

### PR DESCRIPTION
**Pull request by Billccx**
## Motivation
Enable top-down methods to predict keypoints visibility

## Modification
1. add a new head `vis_head`      (mmpose/mmpose/models/heads/vis_heads)
2. add a new TTA function `flip_visibility`     (mmpose/mmpose/models/utils/tta.py)
3. add a new key `keypoints_visible` in `label_mapping_table` (mmpose/mmpose/datasets/transforms/formatting.py)

## Use cases (Optional)

We can use `vis_head` to wrap a keypoints prediction head, getting the keypoints prediction results with keypoints visibility.

For example, in config file:

```
head=dict(
      type='VisibilityPredictionHead',
      in_channels=2048,
      num_joints=17,
      loss=dict(type='BCELoss', use_target_weight=True),
      pose_cfg=dict(
          type='RegressionHead',
          in_channels=2048,
          num_joints=17,
          loss=dict(type='SmoothL1Loss', use_target_weight=True),
          decoder=codec
      )
  )
```

## Checklist

**Before PR**:

- [x] I have read and followed the workflow indicated in the [CONTRIBUTING.md](https://github.com/open-mmlab/mmpose/blob/master/.github/CONTRIBUTING.md) to create this PR.
- [x] Pre-commit or linting tools indicated in [CONTRIBUTING.md](https://github.com/open-mmlab/mmpose/blob/master/.github/CONTRIBUTING.md) are used to fix the potential lint issues.
- [x] Bug fixes are covered by unit tests, the case that causes the bug should be added in the unit tests.
- [x] New functionalities are covered by complete unit tests. If not, please add more unit tests to ensure correctness.
- [x] The documentation has been modified accordingly, including docstring or example tutorials.

**After PR**:

- [x] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: mmpose/models/heads/vis_heads/vis_head.py [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+            feats (Tuple[Tensor]): Multi scale feature maps.
+
+        Returns:
+            Tensor: output coordinates visibility.
+        """
+        x = feats[-1]
+        while len(x.shape) < 4:
+            x.unsqueeze_(-1)
+        x = self.vis_head(x)
+        return x.reshape(-1, self.num_joints)
```
Comment (Tau-J) [inline mmpose/models/heads/vis_heads/vis_head.py:66]: This implementation will cause the `pose_head` to be lost during the export of onnx. Also, the user must manually call `pose_head` for inference, instead of getting all results at once, which is not ideal.
I suggest performing `pose_head` inference in `forward()` and returning results for both `vis_head` and `pose_head` at the same time.

File: mmpose/models/heads/__init__.py [hunk 2]
```diff
@@ -6,9 +6,11 @@
 from .hybrid_heads import DEKRHead
 from .regression_heads import (DSNTHead, IntegralRegressionHead,
                                RegressionHead, RLEHead)
+from .vis_heads import VisibilityPredictionHead
```
Comment (Tau-J) [inline mmpose/models/heads/__init__.py:9]: Head with multiple outputs can be put in `hybrid_heads` folder.

File: mmpose/models/heads/vis_heads/vis_head.py [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+        self.in_channels = in_channels
+        self.num_joints = num_joints
+        self.loss_module = MODELS.build(loss)
+
+        self.pose_head = MODELS.build(pose_cfg)
+        self.pose_cfg = pose_cfg
+
+        self.vis_head = nn.Sequential(
+            nn.AdaptiveAvgPool2d(1), nn.Flatten(),
+            nn.Linear(self.in_channels, self.num_joints), nn.Sigmoid())
```
Comment (Tau-J) [inline mmpose/models/heads/vis_heads/vis_head.py:50]: I suggest adding a `use_sigmoid: bool = False` in `__init__()` to determine wether to use `sigmoid` in `vis_head`, because Pytorch officially recommand to use `BCEWithLogitsLoss` instead of `Sigmoid + BCELoss`, which is "more numerically stable than using a plain Sigmoid followed by a BCELoss as, by combining the operations into one layer, we take advantage of the log-sum-exp trick for numerical stability." [details](https://pytorch.org/docs/stable/generated/torch.nn.BCEWithLogitsLoss.html)

File: mmpose/datasets/transforms/formatting.py [hunk 4]
```diff
@@ -99,7 +99,8 @@ class PackPoseInputs(BaseTransform):
         'keypoint_x_labels': 'keypoint_x_labels',
         'keypoint_y_labels': 'keypoint_y_labels',
         'keypoint_weights': 'keypoint_weights',
-        'instance_coords': 'instance_coords'
+        'instance_coords': 'instance_coords',
+        'keypoints_visible': 'keypoints_visible',
```
Comment (Tau-J) [inline mmpose/datasets/transforms/formatting.py:103]: Could you please support `transformed_keypoints_visible` in https://github.com/open-mmlab/mmpose/blob/0796d77361201a862e9341d512b5445751df8832/mmpose/datasets/transforms/formatting.py#L156 
and https://github.com/open-mmlab/mmpose/blob/0796d77361201a862e9341d512b5445751df8832/mmpose/datasets/transforms/common_transforms.py#L935
because keypoint visibility will be used in `browse_dataset.py`

**Review by Tau-J — CHANGES_REQUESTED** [review]
request changes
